airbyte-cdk/python/airbyte_cdk/sources/abstract_source.py

```diff
@@ -106,6 +107,11 @@ def read(
                         f"The requested stream {configured_stream.stream.name} was not found in the source."
                         f" Available streams: {stream_instances.keys()}"
                     )
+                if self.availability_strategy is not None:
+                    stream_is_available, error = self.availability_strategy.check_availability(logger, stream_instance)
```

Comment:
> I'd be interested to see what it would take to manage availability at the stream level (I think this is an option you're considering).

Yep, looking into that! 

> Moreover, what do you think about building a newly configured catalog from which we've stripped the unavailable streams in a specific availability_check step? The availability check could occur between discover and read.

Not sure I understand here, is the idea to pull the logic completely out of read, and not read sources that we've just previously determined to be available? Is this assuming we run discover and get a new configured catalog before every read, or just the one time when we create the initial catalog for the user to configure?

Reply:
You could change  `read_catalog` [here](https://github.com/airbytehq/airbyte/blob/master/airbyte-cdk/python/airbyte_cdk/sources/source.py#L89)  to return a catalog with only the available streams.  This method is called in the entrypoint [here](https://github.com/airbytehq/airbyte/blob/master/airbyte-cdk/python/airbyte_cdk/entrypoint.py#L119).

Reply:
Great idea, makes sense, will implement in a separate PR! Made a [new issue](https://app.zenhub.com/workspaces/connector-operations-6318ede5eda32e0010ef8566/issues/airbytehq/airbyte/20252) for it